Added api discovery to kubectl run, based on which we decide which generators to use

soltysh opened the pull request:
Fixes #22836.

@bgrant0607 @jlowdermilk @lavalamp @deads2k @soltysh @janetkuo wdyt?

PS If fixes/changes are needed after 4pm UTC take over this PR and resubmit since I'll be out the entire evening.

soltysh committed:
Added api discovery to kubectl run, based on which we decide which generators to use

deads2k commented on pkg/kubectl/cmd/run.go:
```diff
@@ -146,10 +147,27 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
+		}
 		if restartPolicy == api.RestartPolicyAlways {
-			generatorName = "deployment/v1beta1"
+			if contains(groupList, "extensions/v1beta1") {
```
use constants for group/versions and use `GroupVersion` objects.

deads2k replied:
You're going to need some comments explaining priority decisions here.  I know it, but I think its because I just read the issue and a month from now someone is going to be looking through. and wondering.

deads2k commented on pkg/kubectl/cmd/run.go:
```diff
@@ -146,10 +147,27 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
 		if restartPolicy == api.RestartPolicyAlways {
-			generatorName = "deployment/v1beta1"
+			if contains(groupList, "extensions/v1beta1") {
+				generatorName = "deployment/v1beta1"
```
Just because you have the group doesn't mean you enabled the object.  As I recall, we had levels with different resources enabled by default.  How about checking both?

soltysh replied:
True, that is the case with 1.1 I've just setup.

deads2k commented:
If we have a decent mock for the discovery client I'd like to see a unit test or two, but this looks fine.

bgrant0607 commented on pkg/kubectl/cmd/run.go:
```diff
@@ -146,10 +147,27 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
+		groupList, err := client.Discovery().ServerGroups()
+		if err != nil {
+			// this cover the cases where old servers do not expose discovery
+			groupList = nil
```
This is what makes it work against 1.1?

soltysh replied:
That's my idea, I don't have any 1.1 cluster at hand :( Although I can compile 1.1 version and see what I'm gonna get.

bgrant0607 commented:
Thanks for jumping on this so quickly!

Could you please update the description text for --generator?

soltysh commented:
@deads2k @bgrant0607 I've addressed all the comments. Tested that against 1.1 cluster. It nicely falls back all the way back to running pods if jobs are disabled in 1.1 (the default behavior). ptal

bgrant0607 commented:
Thanks, @soltysh.

Please run hack/update-all.sh.

soltysh commented:
@bgrant0607 done, sorry

bgrant0607 commented:
LGTM, assuming tests pass. Thanks much

deads2k commented on pkg/kubectl/cmd/run.go:
```diff
@@ -254,6 +276,22 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
 	return nil
 }
 
+func contains(resourcesList map[string]*unversioned.APIResourceList, gv unversioned.GroupVersion, name string) bool {
```
`GroupVersionResource`?

deads2k replied:
Also, how about a `TODO` to make this more library-esq.  It looks like it would have general utility as a layer on top of a discovery client.

bgrant0607 commented:
Thanks for the comments @deads2k. Removed lgtm. Please reapply when your comments are addressed.

soltysh commented:
On it...

soltysh commented:
@deads2k @bgrant0607 comments addressed ptal.

soltysh commented:
If more tweaks are needed or tests fail please pick up this PR because I'll be leaving and won't be able to touch it and I do realize this is important to have in as soon as possible.

bgrant0607 commented:
Thanks @soltysh

bgrant0607 commented:
There are no unit test results, so I can't see what failed.

@k8s-bot unit test this issue: #IGNORE

deads2k commented:
lgtm

bgrant0607 commented:
All green. Merging @k8s-oncall

eparis commented:
This PR was sucessfully cherry picked in PR #22855
please verify that the release-1.2 branch contains these changes as you would expect and contact @eparis if there appear to be problems.
